### PR TITLE
UnsafeWriter: reuse allocated blocks between threads 

### PR DESCRIPTION
**Pull request by ulex**
Currently unsafe writer is allocated per thread, which leads to increased memory usage in some scenarios.
Proposed PR introduces the NativeMemoryPool concept for reusing blocks of memory between threads. 

Measured difference when simply launching VS + ReSharper and waiting for caches:
Before optimization: 29 MB
After: 9 MB

In addition, there is an API for tracking allocated blocks and freeing unused blocks (both manually and automatically)

I have not been able to find any performance degradation compared to the existing implmentation.

### CODE REVIEW

**Comment by hypersw** [comment]
@korifey : this is adding fencing / core syncing to unsafe writer operations (in the form of Monitor Enter at the least), how would you comment on the impact? I remember this is against the original design, but maybe this were reconsidered or something?

**Comment by ulex** [comment]
@hypersw thanks for review. Let me clarify a little proposed implementation. The lock is only required when there are not enought allocated blocks in the pool or when we want to clear already allocated blocks. One need to allocate new block only when more concurrent UnsafeWriters are required. Clear, currently only happens only when the thread that allocated the block has died (like in existing implementation). Most of the time, allocted block returns to the pool and can be reused lock-free. Rougt perfomance tests were unable to detect a significant difference in highly concurrent multithreading scenario comparing to the existing UnsafeWriter implementation, but it is possible that I am missing something.

**Comment by sergeyQx** [comment]
Looks like number of allocated UnsafeWriters should not be bigger than number of active threads if we free on UnsafeWriter on thread disposal

**Comment by korifey** [comment]
> @korifey : this is adding fencing / core syncing to unsafe writer operations (in the form of Monitor Enter at the least), how would you comment on the impact? I remember this is against the original design, but maybe this were reconsidered or something?

Initial design was optimized for mass UnsafeWriter.Cookie creation and disposal - it was nearly free. So I took into account low-level optimizations like inlining & branch predictions. I guess the only drawback in design proposed by @ulex is 2-CompareExchanges (one on Cookie acquire, one on disposal) and little bit more branching (need to look in profiles). If it's ok, then I see no problems here.

**Comment by ulex** [comment]
@sergeyQx 
> Looks like number of allocated UnsafeWriters should not be bigger than number of active threads if we free on UnsafeWriter on thread disposal

Applying such strategy would lead to nearly-empty pool. Threads die offten and if you just use already allocated block you cannot be responsible for freeing it. On the other hand, when you are lucky enough to trigger an allocation, you have to free allocated block.
I've clarified current memory management and removed an inncacurate hack, thanks for your comment.

File: rd-net/Lifetimes/Serialization/NativeMemoryPool.cs [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+    {
+      // try to search in the array of all memory blocks
+      ThreadMemoryHolder h;
+      if (SearchAndReserve(out h))
+        return new Cookie(h);
+      if (AllocateNew(out h))
+        return new Cookie(h, causedAllocation: true);
+
+      // All items in array blocks are busy and not available, increase the array size and try once more
+      lock (ourLock)
```
[the remark this answers is not in the document]
Comment (ulex) [reply rd-net/Lifetimes/Serialization/NativeMemoryPool.cs:100]: > other threads might jump in and occupy the empty slots
The others threads might occupy the the empty slot (by which I mean change null to non-null for some item in ourBlocks) only in AllocateNew, whose body is also protected by ourLock. I see no problem here.

[the remark this answers is not in the document]
Comment (sergeyQx) [reply rd-net/Lifetimes/Serialization/NativeMemoryPool.cs:100]: You are right. Withdrawn.

**Review by sergeyQx — APPROVED** [review]

**Pull request merged**